File: libraries/sdk_binder/android/src/main/java/com/perplelab/unityads/PerpleUnityAds.java
package com.perplelab.unityads;

import com.perplelab.PerpleSDK;
import com.perplelab.PerpleLog;

import com.unity3d.ads.IUnityAdsListener;
import com.unity3d.ads.UnityAds;
import com.unity3d.ads.UnityAds.FinishState;
import com.unity3d.ads.UnityAds.UnityAdsError;
import com.unity3d.ads.metadata.MediationMetaData;
import com.unity3d.ads.metadata.PlayerMetaData;

import android.app.Activity;
import android.os.Handler;

import org.json.JSONException;
import org.json.JSONObject;

public class PerpleUnityAds implements IUnityAdsListener {
    private static final String LOG_TAG = "PerpleSDK UnityAds";

    private Handler mAppHandler;
    private PerpleUnityAdsCallback mCallback;
    private boolean mIsDebug;
    private String mGameId;

    public PerpleUnityAds() {}

    public void init(String gameId, boolean isDebug) {
        mGameId = gameId;
        mIsDebug = isDebug;
        mAppHandler = new Handler();
    }

    public void start(boolean isTestMode, String metaData, PerpleUnityAdsCallback callback) {
        mCallback = callback;

        if (UnityAds.isInitialized()) {
            if (mCallback != null) {
                mCallback.onError("ALREADY_INITIALIZED", "");
            }
            return;
        }

        if (!metaData.isEmpty()) {
            String name = "";
            String version = "";
            try {
                JSONObject jsonObj = new JSONObject(metaData);
                name = (String)jsonObj.get("name");
                version = (String)jsonObj.get("version");
            } catch (JSONException e) {
                e.printStackTrace();
            }

            MediationMetaData mediationMetaData = new MediationMetaData(PerpleSDK.getInstance().getMainActivity());
            mediationMetaData.setName(name);
            mediationMetaData.setVersion(version);
            mediationMetaData.commit();
        }

        UnityAds.initialize(PerpleSDK.getInstance().getMainActivity(), mGameId, this, isTestMode);
        UnityAds.setDebugMode(mIsDebug);
    }

    public void show(String placementId, String metaData) {
        if (!UnityAds.isInitialized()) {
            PerpleLog.e(LOG_TAG, "UnityAds is not initialized.");
            if (mCallback != null) {
                mCallback.onError("NOT_INITIALIZED", "");
            }
            return;
        }

        String serverId = "";
        String ordinalId = "";
        if (!metaData.isEmpty()) {
            try {
                JSONObject jsonObj = new JSONObject(metaData);
                serverId = (String)jsonObj.get("serverId");
                ordinalId = (String)jsonObj.get("ordinalId");
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        Activity showActivity = PerpleSDK.getInstance().getMainActivity();
        if (placementId.isEmpty()) {
            if (UnityAds.isReady()) {

                if (serverId != null && !serverId.isEmpty()) {
                    PlayerMetaData playerMetaData = new PlayerMetaData(showActivity);
                    playerMetaData.setServerId(serverId);
                    playerMetaData.commit();
                }

                if (ordinalId != null && !ordinalId.isEmpty()) {
                    MediationMetaData mediationMetaData = new MediationMetaData(showActivity);
                    mediationMetaData.setOrdinal(Integer.parseInt(ordinalId));
                    mediationMetaData.commit();
                }

                UnityAds.show(showActivity);
            } else {
                if (mCallback != null) {
                    mCallback.onError("NOT_READY", "");
                }
            }
        } else {
            if (UnityAds.isReady(placementId)) {

                if (serverId != null && !serverId.isEmpty()) {
                    PlayerMetaData playerMetaData = new PlayerMetaData(showActivity);
                    playerMetaData.setServerId(serverId);
                    playerMetaData.commit();
                }

                if (ordinalId != null && !ordinalId.isEmpty()) {
                    MediationMetaData mediationMetaData = new MediationMetaData(showActivity);
                    mediationMetaData.setOrdinal(Integer.parseInt(ordinalId));
                    mediationMetaData.commit();
                }

                UnityAds.show(showActivity, placementId);
            } else {
                if (mCallback != null) {
                    mCallback.onError("NOT_READY", "");
                }
            }
        }
    }

    @Override
    public void onUnityAdsError(final UnityAdsError error, final String message) {
        mAppHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mCallback != null) {
                    mCallback.onError(error.toString(), message);
                }
            }
        });
    }

    @Override
    public void onUnityAdsFinish(final String placementId, final FinishState result) {
        mAppHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mCallback != null) {
                    mCallback.onFinish(placementId, result.toString());
                }
            }
        });
    }

    @Override
    public void onUnityAdsReady(final String placementId) {
        mAppHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mCallback != null) {
                    mCallback.onReady(placementId);
                }
            }
        });
    }

    @Override
    public void onUnityAdsStart(final String placementId) {
        mAppHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mCallback != null) {
                    mCallback.onStart(placementId);
                }
            }
        });
    }
}

File: libraries/sdk_binder/iOS/ios/PerpleUnityAds.m
//
//  PerpleUnityAds.m
//  PerpleSDK
//
//  Created by PerpleLab on 2017. 8. 16..
//  Copyright © 2017년 PerpleLab. All rights reserved.
//
/*
#import "PerpleUnityAds.h"

@implementation PerpleUnityAds

@synthesize mParentView;
@synthesize mGameId;
@synthesize mIsDebug;
@synthesize mCallback;

- (id) initWithGameId:(NSString *)gameId
           parentView:(UIViewController *)view
                debug:(BOOL)isDebug {

    if (self = [super init]) {
        self.mGameId = gameId;
        self.mIsDebug = isDebug;
        self.mParentView = view;
    } else {
        NSLog(@"PerpleUnityAds, UnityAds initializing fail.");
    }

    return self;
}

- (void) dealloc {
    self.mParentView = nil;
    self.mGameId = nil;

#if !__has_feature(objc_arc)
    [super dealloc];
#endif
}

- (void)start:(BOOL)isTestMode
     metaData:(NSString *)metaData
   completion:(PerpleSDKCallback)callback {
    self.mCallback = callback;

    if ([UnityAds isInitialized]) {
        if (self.mCallback) {
            self.mCallback(@"error", @"ALREADY_INITIALIZED");
        }
        return;
    }

    if (![metaData isEqualToString:@""]) {
        NSDictionary *dic = [PerpleSDK getNSDictionaryFromJSONString:metaData];
        NSString *name = [dic objectForKey:@"name"];
        NSString *version = [dic objectForKey:@"version"];

        id mediationMetaData = [[UADSMediationMetaData alloc] init];
        [mediationMetaData setName:name];
        [mediationMetaData setVersion:version];
        [mediationMetaData commit];
    }

    [UnityAds initialize:self.mGameId testMode:isTestMode];
    [UnityAds addDelegate:self];
    [UnityAds setDebugMode:self.mIsDebug];
}

- (void)show:(NSString *)placementId
    metaData:(NSString *)metaData {

    if (![UnityAds isInitialized]) {
        if (self.mCallback) {
            self.mCallback(@"error", @"NOT_INITIALIZED");
        }
        return;
    }

    NSString *serverId = @"";
    NSString *ordinalId = @"";

    if (![metaData isEqualToString:@""]) {
        NSDictionary *dic = [PerpleSDK getNSDictionaryFromJSONString:metaData];
        serverId = [dic objectForKey:@"serverId"];
        ordinalId = [dic objectForKey:@"ordinalId"];
    }

    if ([placementId isEqualToString:@""]) {
        if ([UnityAds isReady]) {

            if (serverId != nil && ![serverId isEqualToString:@""]) {
                id playerMetaData = [[UADSPlayerMetaData alloc] init];
                [playerMetaData setServerId:serverId];
                [playerMetaData commit];
            }

            if (ordinalId != nil && ![ordinalId isEqualToString:@""]) {
                id mediationMetaData = [[UADSMediationMetaData alloc] init];
                [mediationMetaData setOrdinal:[ordinalId intValue]];
                [mediationMetaData commit];
            }

            [UnityAds show:self.mParentView];
        } else {
            if (self.mCallback) {
                self.mCallback(@"error", @"NOT_READY");
            }
        }

    } else {
        if ([UnityAds isReady:placementId]) {

            if (serverId != nil && ![serverId isEqualToString:@""]) {
                id playerMetaData = [[UADSPlayerMetaData alloc] init];
                [playerMetaData setServerId:serverId];
                [playerMetaData commit];
            }

            if (ordinalId != nil && ![ordinalId isEqualToString:@""]) {
                id mediationMetaData = [[UADSMediationMetaData alloc] init];
                [mediationMetaData setOrdinal:[ordinalId intValue]];
                [mediationMetaData commit];
            }

            [UnityAds show:self.mParentView placementId:placementId];
        } else {
            if (self.mCallback) {
                self.mCallback(@"error", @"NOT_READY");
            }
        }
    }
}

- (void)unityAdsReady:(NSString *)placementId {
    if (self.mCallback) {
        self.mCallback(@"ready", placementId);
    }
}

- (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
    if (self.mCallback) {
        NSString *errorCode = @"ERROR";
        switch (error) {
            case kUnityAdsErrorInitializedFailed:
                errorCode = @"INITIALIZE_FAILED";
                break;
            case kUnityAdsErrorVideoPlayerError:
                errorCode = @"VIDEO_PLAYER_ERROR";
                break;
            case kUnityAdsErrorNotInitialized:
                errorCode = @"NOT_INITIALIZED";
                break;
            case kUnityAdsErrorDeviceIdError:
                errorCode = @"DEVICE_ID_ERROR";
                break;
            case kUnityAdsErrorShowError:
                errorCode = @"SHOW_ERROR";
                break;
            case kUnityAdsErrorFileIoError:
                errorCode = @"FILE_IO_ERROR";
                break;
            case kUnityAdsErrorInternalError:
                errorCode = @"INTERNAL_ERROR";
                break;
            case kUnityAdsErrorInvalidArgument:
                errorCode = @"INVALID_ARGUMENT";
                break;
            case kUnityAdsErrorAdBlockerDetected:
                errorCode = @"AD_BLOCKER_DETECTED";
                break;
            case kUnityAdsErrorInitSanityCheckFail:
                errorCode = @"INIT_SANITY_CHECK_FAIL";
                break;
        }
        self.mCallback(@"error", errorCode);
    }
}

- (void)unityAdsDidStart:(NSString *)placementId {
    if (self.mCallback) {
        self.mCallback(@"start", placementId);
    }
}

- (void)unityAdsDidFinish:(NSString *)placementId withFinishState:(UnityAdsFinishState)state {
    if (self.mCallback) {
        NSString *result = @"ERROR";
        switch (state) {
            case kUnityAdsFinishStateError:
                result = @"ERROR";
                break;
            case kUnityAdsFinishStateSkipped:
                result = @"SKIPPED";
                break;
            case kUnityAdsFinishStateCompleted:
                result = @"COMPLETED";
                break;
        }
        NSString *info = [PerpleSDK getJSONStringFromNSDictionary:@{@"placementId":placementId, @"result":result}];
        self.mCallback(@"finish", info);
    }
}

@end
*/
